File: edukate-checker/src/main/java/io/github/sanyavertolet/edukate/checker/configs/RestClientConfig.java
package io.github.sanyavertolet.edukate.checker.configs;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.http.client.ClientHttpRequestFactoryBuilder;
import org.springframework.boot.http.client.ClientHttpRequestFactorySettings;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.web.client.RestClient;

import java.time.Duration;

@Configuration
public class RestClientConfig {

    /**
     *  Spring AI uses RestClient.Builder to build a model client but provides no way to configure timeouts.
     */
    @Bean
    public RestClient.Builder restClientBuilder(
            @Value("${spring.http.client.connect-timeout:1s}") Duration connectTimeout,
            @Value("${spring.http.client.read-timeout:10s}") Duration readTimeout
    ) {
        ClientHttpRequestFactorySettings settings = ClientHttpRequestFactorySettings.defaults()
                .withConnectTimeout(connectTimeout)
                .withReadTimeout(readTimeout);
        return RestClient.builder().requestFactory(ClientHttpRequestFactoryBuilder.reactor().build(settings));
    }
}

File: edukate-checker/src/main/kotlin/io/github/sanyavertolet/edukate/checker/configs/RestClientConfig.kt
package io.github.sanyavertolet.edukate.checker.configs

import java.time.Duration
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.http.client.ClientHttpRequestFactoryBuilder
import org.springframework.boot.http.client.HttpClientSettings
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.web.client.RestClient

@Configuration
class RestClientConfig {
    /** Spring AI uses RestClient.Builder to build a model client but provides no way to configure timeouts. */
    @Bean
    fun restClientBuilder(
        @Value($$"${spring.http.clients.connect-timeout:5s}") connectTimeout: Duration,
        @Value($$"${spring.http.clients.read-timeout:2m}") readTimeout: Duration,
    ): RestClient.Builder {
        val settings = HttpClientSettings.defaults().withConnectTimeout(connectTimeout).withReadTimeout(readTimeout)
        return RestClient.builder().requestFactory(ClientHttpRequestFactoryBuilder.reactor().build(settings))
    }
}
